Add upgradable guard to RwLock

ohanar opened the pull request:
This addresses #21.

Amanieu commented:
Thanks, this looks great!

This is quite a lot of code, so I'll review it over the week-end. From a quick look so far, it looks mostly good though!

Amanieu commented on src/raw_rwlock.rs:
```diff
@@ -13,19 +13,53 @@ use stable::{AtomicUsize, Ordering};
+    #[cfg(not(feature = "nightly"))] {
+        !0
+    }
+};
```
What is the purpose of this constant? It seems that it always has the value `!0`.

ohanar replied:
I personally need to include this in a more complicated data structure (tagged pointers and the like...), so I wanted to make sure the implementation works and doesn't panic when using only a few bits for the state.

I added this so that it was easy for me to switch between testing with just a few bits and when using the full `usize`. In particular, everything should work perfectly fine with a value of `0b1111`, just with some different performance characteristics.

ohanar commented:
Yes, it definitely needs a look over as it is very tricky code. (and probably a few more tests wouldn't hurt)

ohanar commented on src/raw_rwlock.rs:
```diff
@@ -13,19 +13,53 @@ use stable::{AtomicUsize, Ordering};
+// An exclusive lock acquires all of guard resource (i.e. it is exclusive)
+const EXCLUSIVE_GUARD: usize = GUARD_COUNT_MASK;
+// An upgradable lock acquires just over half of the guard resource
+const UPGRADABLE_GUARD: usize = (GUARD_COUNT_MASK >> 1) + (SHARED_GUARD >> 1);
```
It's currently correct. It should be `(GUARD_COUNT_MASK + SHARED_GUARD) >> 1` (i.e. just over half of the guards), but that addition might overflow. I'll add a comment to the source.

ohanar commented:
I've hopefully addressed all of your concerns, let me know if you find anything else concerning. I would be happy to squash the commits before merging.

Amanieu commented:
Well, I can't find anything wrong with the implementation, so I'll merge it. Thanks for the PR!

Amanieu commented:
Published in v0.5.2
